File: backend/src/main/java/com/lsoft/aguapotable/controller/PagoController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.domain.entity.Pago;
import com.lsoft.aguapotable.domain.entity.Usuario;
import com.lsoft.aguapotable.domain.repository.PagoRepository;
import com.lsoft.aguapotable.domain.repository.UsuarioRepository;
import com.lsoft.aguapotable.dto.request.PagoRequest;
import com.lsoft.aguapotable.service.PagoService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/pagos")
@RequiredArgsConstructor
@Tag(name = "Pagos", description = "Registro y consulta de pagos")
@SecurityRequirement(name = "bearerAuth")
public class PagoController {

    private final PagoService pagoService;
    private final PagoRepository pagoRepository;
    private final UsuarioRepository usuarioRepository;

    @PostMapping
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN','CAJERO')")
    @Operation(summary = "Registrar un pago sobre una factura")
    public ResponseEntity<Pago> registrar(@Valid @RequestBody PagoRequest dto,
                                           Authentication authentication) {
        Usuario cajero = usuarioRepository.findByEmail(authentication.getName())
                .orElse(null);

        Pago pago = pagoService.registrarPago(
                dto.getFacturaId(),
                dto.getMonto(),
                dto.getMetodoPago(),
                dto.getReferencia(),
                cajero
        );
        return ResponseEntity.status(201).body(pago);
    }

    @GetMapping("/factura/{facturaId}")
    @Operation(summary = "Listar pagos de una factura")
    public ResponseEntity<List<Pago>> getPorFactura(@PathVariable Long facturaId) {
        return ResponseEntity.ok(pagoRepository.findByFacturaId(facturaId));
    }
}

File: backend/src/pago/pago.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Param,
  ParseIntPipe,
  HttpCode,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { PagoService } from './pago.service';
import { PagoRequest } from './dto/pago.dto';
import { Usuario } from '../common/entities';
import { CurrentUser } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Pagos')
@ApiBearerAuth()
@Controller('api/v1/pagos')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class PagoController {
  constructor(
    private readonly pagoService: PagoService,
    @InjectRepository(Usuario)
    private readonly usuarioRepo: Repository<Usuario>,
  ) {}

  @Post()
  @Roles('SUPER_ADMIN', 'ADMIN', 'CAJERO')
  @HttpCode(201)
  @ApiOperation({ summary: 'Registrar un pago sobre una factura' })
  async registrar(
    @Body() dto: PagoRequest,
    @CurrentUser() user: any,
  ) {
    const cajero = await this.usuarioRepo.findOne({
      where: { email: user.email },
    });
    return this.pagoService.registrarPago(
      dto.facturaId,
      dto.monto,
      dto.metodoPago!,
      dto.referencia,
      cajero,
    );
  }

  @Get('factura/:facturaId')
  @ApiOperation({ summary: 'Listar pagos de una factura' })
  async getPorFactura(@Param('facturaId', ParseIntPipe) facturaId: number) {
    return this.pagoService.getPorFactura(facturaId);
  }
}
